### ISSUE TITLE
breakable tcolorbox may overlap with following text

### ISSUE THREAD

**Issue by muzimuzhi** [T-F-S/tcolorbox#106]
From TeX-SX: https://tex.stackexchange.com/q/551112

The following example is slightly different from the one found in TeX-SX question. To emphasis the overlapping result, text after `tcolorbox` is in blue.
```tex
\documentclass{article}
\usepackage[breakable]{tcolorbox}
\usepackage{lipsum}

\begin{document}
\lipsum[1-4]

\begin{tcolorbox}[size=minimal,breakable]
  \begin{enumerate}
    \item
    \lipsum[1-3]
    \vskip 50pt plus 0pt minus 50pt
    \lipsum[1-2]
  \end{enumerate}
\end{tcolorbox}

\color{blue}\lipsum[1]
\end{document}
```
![image](https://user-images.githubusercontent.com/6376638/85931351-6558b780-b8f6-11ea-93e2-06c8390613b1.png)

My knowledge about TeX's page breaking scheme is limited, hence the following info may not be helpful.

Copied from my comment under that question:

> I guess the problem is caused by too much shrinkable space in `tcolorbox`. `longtable` package, which constructs a long and breakable tabular, encounters similar problem, and a workaround/solution for `longtable` given by David Carlisle can be seen [here](https://github.com/latex3/latex2e/issues/183#issuecomment-596204131).

Diff results of David Carlisle's workaround/solution for `longtable`:
```diff
diff --git a/diff-old.tex b/diff-new.tex
index cbc7123..b061231 100644
--- a/diff-old.tex
+++ b/diff-new.tex
@@ -1,6 +1,13 @@
 \def\LT@start{%
   \let\LT@start\endgraf
   \endgraf\penalty\z@\vskip\LTpre\endgraf
+% new test from Lars Hellström force TeX to consider a page break here
+% https://www.latex-project.org/cgi-bin/ltxbugs2html?pr=tools/3396
+   \ifdim \pagetotal<\pagegoal \else
+      \dimen@=\pageshrink
+      \advance \dimen@ 1sp
+      \kern\dimen@\penalty 9999\endgraf \kern-\dimen@
+   \fi
   \dimen@\pagetotal
   \advance\dimen@ \ht\ifvoid\LT@firsthead\LT@head\else\LT@firsthead\fi
   \advance\dimen@ \dp\ifvoid\LT@firsthead\LT@head\else\LT@firsthead\fi
@@ -16,10 +23,12 @@
   \advance\dimen@\dp
         \ifdim\dp\@arstrutbox>\dp\tw@\@arstrutbox\else\tw@\fi
   \advance\dimen@ -\pagegoal
+% new test from me zero out pageshrink as page-so-far is boxed
+\ifdim\pageshrink>\z@\pageshrink\z@\fi
   \ifdim \dimen@>\z@\vfil\break\fi
       \global\@colroom\@colht
   \ifvoid\LT@foot\else
-    \advance\vsize-\ht\LT@foot
+    \global\advance\vsize-\ht\LT@foot
     \global\advance\@colroom-\ht\LT@foot
     \dimen@\pagegoal\advance\dimen@-\ht\LT@foot\pagegoal\dimen@
     \maxdepth\z@
```

**Comment by T-F-S** [comment]
Thank your for reporting. I copy my answer from https://tex.stackexchange.com/q/551112

Suprisingly, I have not encountered this situation before.

As commented by you, there seems to be to much shrinkable space inside the box. The break algorithm detects that the box cannot be broken further, but the resulting last box seems to be too large to fit on the page (not really true here). So, the bounding box for the last box part is made smaller to fit on the page (we would get a nasty empty page otherwise).

Here, the reduction is way too large leading to the seen overlap.

Preferably, the break algorithm should be more cunning (?!), but at least I can offer a solution which inserts a `\clearpage` automatically:

```latex
    \makeatletter
    \def\tcb@split@force@last{%
      \tcb@split@setstate@last%
      \ifdim\tcb@h@total>\tcb@h@page\relax%
        \gdef\tcb@after@lastbox{\clearpage}%
        \tcbdimto\kvtcb@bbbottom{\kvtcb@bbbottom+\tcb@h@page-\tcb@h@total}%
      \fi%
    }
    \makeatother
```

This patch should do the trick. The only (minor) drawback is that it removes user code given by the `after` key. But, IMHO this key is intended for spacing after the box and is not relevant for a page break.

I hesitate very much to change anything about the break algorithm, because any change typically causes a lot of trouble in existing documents. But I think this change is useful and should give not too much harm.

**Comment by muzimuzhi** [comment]
Many thanks.

**Comment by muzimuzhi** [comment]
Note: resolved in v4.31 by https://github.com/T-F-S/tcolorbox/commit/55d9deed444eef58c65a88b893500255f44ac7ba, more specifically by [these lines](https://github.com/T-F-S/tcolorbox/commit/55d9deed444eef58c65a88b893500255f44ac7ba#diff-44b21426c05f1ac0c8894ba07f563172R639-R640).
